Support overriding timeout through by environment variable.

lorenzodalaqua opened the pull request:
For https://github.com/Medology/Umami/issues/2823

This PR adds logic for overriding the `timeout` extension configuration option by setting a value on an environment valuable called `BEHAT_RETRY_TIMEOUT`.

The extension will pull the configuration value from the env variable first, then `behat.yml`, and it can be used to set timeouts per-run with `BEHAT_RETRY_TIMEOUT=X behat ...`.

Note: for projects that use a containerized PHP, consumer will need to ensure the host environment variable is passed to the container on the PHP script as we did for the container on this project.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * You can override the retry timeout using BEHAT_RETRY_TIMEOUT; numeric values take effect and invalid/negative values are ignored with a warning.

* **Documentation**
  * Added guidance and an example showing how to use BEHAT_RETRY_TIMEOUT to override the timeout.

* **Tests**
  * Added scenarios and steps to validate env-var overrides, numeric validation, non-negative checks; tests set and clear the variable per scenario.

* **Chores**
  * Docker run now forwards BEHAT_RETRY_TIMEOUT into the container; CI image and composer invocation updated; added composer install step.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lorenzodalaqua committed:
Add environment variable support for retry timeout configuration

This is not tested yet due to requiring changes to a 3rd-party dependency to allow us to test it (or using a different package for testing). This PR aims to enable the functionality and handle coordinating with the dependency maintainers on a follow-up issue (https://github.com/Medology/Umami/issues/2862)

lorenzodalaqua committed:
Clarify usage of BEHAT_RETRY_TIMEOUT environment variable in README

lorenzodalaqua committed:
Fix composer dependencies issues

The current composer.lock causes an error "Your lock file does not contain a compatible set of packages" when installing composer dependencies.

Since this is a package and not an application, composer.lock doesn't make sense. Ideally we would test the package with the lowest and highest version of all supported dependencies so we confirm it fully supports the range in composer.json, but doing that would be out of the scope of this PR.

I am not even sure CI is currently fully working for this package, but all tests pass locally.

lorenzodalaqua committed:
Add automated testing to environment variable override.

This adds a test scenario to confirm the new functionality indeed overrides the configuration value.

If we expose the BEHAT_RETRY_TIMEOUT environment variable on the PHP container it is then available to the Behat child process run on the test.

For this reason, I added a step specifically for setting this environment variable, as that is the only one exposed to the Behat process currently by being proxied by the PHP container to the behat child process, and have added a BeforeScenario hook for clearing only this specific environment variable on the behat parent process to isolate tests.

lorenzodalaqua committed:
CodeRabbit CR: improve validation and add testing

lorenzodalaqua committed:
Add extra scenario to validate environment variable overrides both if higher and lower than timeout

Chekote reviewed (changes requested):
CircleCI didn't run the workflow because this is the first outside PR I've received and CircleCI wasn't configured to run the workflow on PRs from forks. That's been fixed now and the workflow is failing.

I ran workflow on master to confirm none of the failures exists there, and it is passing: https://app.circleci.com/pipelines/github/Chekote/BehatRetryExtension/27/workflows/c7243b8e-d0b6-4ad2-be0e-ca3c624eeb5e

lorenzodalaqua committed:
Use bin/php to run tests

To share the environment variable with the Behat run in the child process, we need to use bin/php

lorenzodalaqua committed:
Update CircleCI configuration to use chekote/php image to match local environment

lorenzodalaqua committed:
Install composer on CircleCI job

lorenzodalaqua committed:
Use machine executor on CircleCI

lorenzodalaqua committed:
Use bin/php to invoke composer

Chekote commented on .circleci/config.yml:
```diff
@@ -34,4 +34,4 @@ jobs:
       # run tests!
-      - run: ./vendor/bin/behat
\ No newline at end of file
+      - run: ./bin/php ./vendor/bin/behat
```
This PR also switches from the `docker` executor to the `machine` executor, which does have Docker available.
